src/evo/mnauth.cpp

```diff
@@ -0,0 +1,141 @@
+// Copyright (c) 2019 The Dash Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "mnauth.h"
+
+#include "activemasternode.h"
+#include "evo/deterministicmns.h"
+#include "masternode-sync.h"
+#include "net.h"
+#include "net_processing.h"
+#include "netmessagemaker.h"
+#include "validation.h"
+
+#include <unordered_set>
+
+void CMNAuth::PushMNAUTH(CNode* pnode, CConnman& connman)
+{
+    if (!fMasternodeMode || activeMasternodeInfo.proTxHash.IsNull()) {
+        return;
+    }
+
+    uint256 signHash;
+    {
+        LOCK(pnode->cs_mnauth);
+        if (pnode->receivedMNAuthChallenge.IsNull()) {
+            return;
+        }
+        // We include fInbound in signHash to forbid interchanging of challenges by a man in the middle. This way
+        // we protect ourself against MITM in this form:
+        //   node1 <- Eve -> node2
+        // It does not protect against:
+        //   node1 -> Eve -> node2
+        // This is ok as we only use MNAUTH as a DoS protection and not for sensitive stuff
+        signHash = ::SerializeHash(std::make_tuple(*activeMasternodeInfo.blsPubKeyOperator, pnode->receivedMNAuthChallenge, pnode->fInbound));
+    }
+
+    CMNAuth mnauth;
+    mnauth.proRegTxHash = activeMasternodeInfo.proTxHash;
+    mnauth.sig = activeMasternodeInfo.blsKeyOperator->Sign(signHash);
+
+    LogPrint("net", "CMNAuth::%s -- Sending MNAUTH, peer=%d\n", __func__, pnode->id);
+
+    connman.PushMessage(pnode, CNetMsgMaker(pnode->GetSendVersion()).Make(NetMsgType::MNAUTH, mnauth));
+}
+
+void CMNAuth::ProcessMessage(CNode* pnode, const std::string& strCommand, CDataStream& vRecv, CConnman& connman)
+{
+    if (!fMasternodeMode) {
+        return;
+    }
+
+    if (!masternodeSync.IsBlockchainSynced()) {
+        // we can't really verify MNAUTH messages when we don't have the latest MN list
+        return;
+    }
+
+    if (strCommand == NetMsgType::MNAUTH) {
```

Comment:
why isn't it assumed this is the case?

Reply:
Nope, it's called for every incoming message. This follows the convention that we use for all Dash specific message handling. For example, everything related to DKGs is handled in `CDKGSessionManager::ProcessMessage` and it by itself checks which message it actually is. This might look a bit strange in the cases where it's only one message that we're interested in, but I prefer to use the same style/convention for all cases.

Reply:
:+1: